### ISSUE TITLE
[2.7] Order version updated whenever order's save method is called

### ISSUE THREAD

**Issue by thenbrent** [woocommerce/woocommerce#13381]
In WC prior to 2.7, an order's version (i.e. the data stored in post meta with meta key `'_order_version'`) would be set only when the order was created.

In WC 2.7+, the version if set in `Abstract_WC_Order_Data_Store_CPT::update()` whenever an existing order is saved (it's important to note, it doesn't even need to have data actually differ and be updated, whenever the `save()` method is called, the order's version will be set and saved). That means the order's version will be whatever version of WC was active the last time the order's `save()` method is called. That is substantially different to the version meta data referring to the version of WC that was active when the order was created.

Apart from introducing a breaking change, this new approach also in many ways makes the version value serve no purpose, or worse, create issues.

For example, if an order is created and paid with WC 2.6, then its date will be updated when it is paid (see #13380). With WC 2.7, that is no longer the case, instead the order's paid date will be updated and it needs to be used to determine when it was paid. The order's version would be the perfect way to identify which logic to use to determine the paid date for an order. However, because the version may be updated to the latest version, it's not possible to rely on that.

**Comment by thenbrent** [comment]
Follow up from @mikejolley via Slack:

> Order version much like product version tracks what version the data is at. Saving the object (via wp admin in 2.6 or via crud in 2.6) SHOULD update the version.

There are two problems with this:

1. the data isn't necessarily at the new version. The `$order->save()` method has been called, but not all data has been updated to whatever the current versions data structure is. It's even possible for no data to have been updated and the version will still be incremented.
2. this wasn't how the order version was treated prior to 2.7. In almost all cases, it was just the version of WC active at the time it was saved. The only time it was updated is when the line item data actually was updated via the admin page.

To address both of these, you could use two different props, one to refer to the version when the order was created (equivalent to the existing order version meta data), and one to refer to the WC version the last time the order was saved (if that's needed). But if changing the existing prop to serve a new purpose, those issues are present.

> #13381 is another example of using something (it sounds like) being used how it was not intended to be used.

The `'_order_version'` was first introduced in 2.3.7 with SHA: 70d125ae. It was used in that commit to figure out how discount totals should be determined with regard to taxes (which were stored differently depending on _which version of WC was used to create the order_ or _save the line items via the admin_). Ever since, it's been possible via the REST API or custom code using publicly accessible API functions to update a lot of different types of data on an order, including tax line items, without the WC version being updated.

I can't see any proof of the intent for the version to always have been updated. I also can't see how developers using the order version would infer that intent based on the available code, documentation and commit history.

**Comment by mikejolley** [comment]
I think you're mistaken here. Take a look at `wc_save_order_items` in 2.6:

https://github.com/woocommerce/woocommerce/blob/release/2.6/includes/admin/wc-admin-functions.php#L378

This is ran on save (in admin). The actual meta is stored so we know what version the order was last saved/created under, so if we ever need to update something to the current version, we know which orders to target.

Running this on `save()` in 2.7 is most appropriate. This matches the product's also which have a version.

> it's important to note, it doesn't even need to have data actually differ and be updated

If there were something to update to the current version, this would need to happen inside the data store, during that save. We've not had to use it yet, but it's there for when needed.

**Comment by mikejolley** [comment]
> The only time it was updated is when the line item data actually was updated via the admin page.

This runs on _every_ save in 2.6.x admin. It's part of the meta boxes.

**Comment by thenbrent** [comment]
> This runs on every save in 2.6.x admin. It's part of the meta boxes.

I'm aware there was one case where the version was updated. I'm saying it was not updated under every single other possible way order data could be modified (e.g. CLI updates, REST API updates, public API functions (e.g. `$order->add_product()` etc.) with WC < 2.7, and now it is with WC 2.7+.

> Running this on save() in 2.7 is most appropriate. This matches the product's also which have a version.

I'm not saying having a piece of data to reflect the updated version isn't useful. I'm trying to point out that repurposing an existing piece of data breaks backward compatibility with the way that data was used previously.

**Comment by mikejolley** [comment]
I disagree that it's being repurposed. It's being consistently updated in 2.7.x to meet it's intended purpose, now the API/CLI and Admin share similar functions.

**Comment by thenbrent** [comment]
OK I think I understand what you're saying now. The order version in WC < 2.7 was updated only when the structure of the data storage was changed. The structure of the data storage only changed once, with 2.3.7, to change the cart discount totals, so the order version almost never needed to be updated.

In 2.7+, the order version is updated anytime the data is saved. Regardless of whether the data structure is changed or not, but it does mean the data will always be in the structure of the order version set on it.

The behaviour is inconsistent between 2.7 and previous versions, but it probably won't create a breaking change, as long as the data is stored in a way that is compatible with 2.3.7 and 2.7.x correctly.

It does look like there is a breakage for orders created before 2.3.7, but I've opened #13383 for that and will close this.

### Summary of Change

This is a verbose summary for posterity to help anyone else the difference:

With 2.3.7-2.6.x: 

1. changing order line items via the admin interface would update them to the WC > 2.3.7 structure (in `wc_save_order_items()`), because of this, it _would_ update the order version. This is the only time the version data is actually different between WC < 2.3.7 and 2.3.7-2.6.14 so it's the only time the order version if updated.
1. changing order line items via any other method, be it the REST API, CLI or another API function (e.g. `WC_Abstract_Order::add_tax()`) would not definitively update the structure to the WC > 2.3.7 structure, so it _would not_ update the order version. The calling code could update them to the newer structure, and if it chose to, it would be responsible for also updating the order version.
1. changing any other data via any other method won't update the order version, because nothing has changed in terms of how the data is stored.

tl;dr - the order version on orders with WC < 2.7 almost never changed, because the data structure almost never changed.

With 2.7.x: 

1. changing order line items via the admin interface would update them to the WC > 2.3.7 structure so it _would_ update the order version (even if there are no changes to the structure).
1. changing order line items via any other method, be it the REST API, CLI or another API function would update the structure to the WC > 2.7.x structure, so it _would_ update the order version  (even if there are no changes to the structure).
1. changing any other data via any other method would update the order version, even if nothing has changed in terms of how the order is being stored, but if something has changed, it will be updated.

tl;dr - the order version on orders with WC 2.7+ will always change, even if the data structure hasn't changed, but the order version will also always reflect the version of the structure.

**Comment by mikejolley** [comment]
Thanks @thenbrent, I copied that into https://github.com/woocommerce/woocommerce/wiki/2.6.x-to-2.7.x-Developer-Migration-Notes#order-version-property
